core/drivers/crypto/caam/hash/caam_hash.c

```diff
@@ -0,0 +1,636 @@
+{
+	TEE_Result ret = CAAM_OUT_MEMORY;
+
+	HASH_TRACE("Allocate Context (0x%p)", ctx);
+
+	/* Allocate the descriptor */
+	ctx->descriptor = caam_calloc_desc(MAX_DESC_ENTRIES);
+	if (!ctx->descriptor) {
+		HASH_TRACE("Allocation descriptor error");
+		goto exit_alloc;
+	}
+
+	/* Initialize the block buffer */
+	ctx->blockbuf.filled = 0;
+	ctx->blockbuf.max = ctx->alg->size_block;
+
+	/* Allocate the CAAM Context register */
+	ret = caam_calloc_align_buf(&ctx->ctx, ctx->alg->size_ctx);
+
+	if (ret != CAAM_NO_ERROR) {
+		HASH_TRACE("Allocation context register error");
+		goto exit_alloc;
+	}
+
+	cache_operation(TEE_CACHEFLUSH, ctx->ctx.data, ctx->ctx.length);
+
+	/* Ensure buffer length is 0 */
+	ctx->ctx.length = 0;
+
+exit_alloc:
+	if (ret != CAAM_NO_ERROR) {
+		/* Free all data allocated */
+		do_free_intern(ctx);
+	}
+
+	return ret;
+}
+
+/*
+ * Free the SW hashing data context
+ *
+ * @ctx    [in/out] Caller context variable
+ */
+static void do_free(struct crypto_hash_ctx *ctx)
+{
+	struct crypto_hash *hash = to_hash_ctx(ctx);
+
+	HASH_TRACE("Free Context (0x%p)", hash->ctx);
+
+	if (hash->ctx) {
+		do_free_intern(hash->ctx);
+		caam_free(hash->ctx);
+	}
+
+	free(hash);
+}
+
+/*
+ * Initialization of the Hash operation
+ *
+ * @ctx   Operation Software context
+ */
+static TEE_Result do_init(struct crypto_hash_ctx *ctx)
+{
+	struct crypto_hash *hash = to_hash_ctx(ctx);
+	struct hashdata *hashdata = hash->ctx;
+
+	HASH_TRACE("Hash Init (0x%p)", hashdata);
+	if (!hashdata)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	/* Initialize the block buffer */
+	hashdata->blockbuf.filled = 0;
+
+	/* Ensure Context length is 0 */
+	hashdata->ctx.length = 0;
+
+	return TEE_SUCCESS;
+}
+
+/*
+ * Update the Hash operation
+ *
+ * @ctx   Operation Software context
+ * @data  Data to hash
+ * @len   Data length
+ */
+static TEE_Result do_update(struct crypto_hash_ctx *ctx, const uint8_t *data,
+			    size_t len)
+{
+	TEE_Result ret = TEE_ERROR_GENERIC;
+	enum caam_status retstatus = CAAM_FAILURE;
+	struct crypto_hash *hash = to_hash_ctx(ctx);
+	struct hashdata *hashdata = hash->ctx;
+	const struct hashalg *alg = NULL;
+	struct caam_jobctx jobctx = {};
+	uint32_t *desc = NULL;
+	size_t fullsize = 0;
+	size_t size_topost = 0;
+	size_t size_todo = 0;
+	size_t size_inmade = 0;
+	size_t inlen = 0;
+	paddr_t paddr_data = 0;
+
+	HASH_TRACE("Hash Update (0x%p)", hashdata);
+
+	if ((!data && len) || !hashdata)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	alg = hashdata->alg;
+
+	if (data) {
+		paddr_data = virt_to_phys((void *)data);
+		if (!paddr_data) {
+			HASH_TRACE("Bad input data virtual address");
+			ret = TEE_ERROR_BAD_PARAMETERS;
+			goto exit_update;
+		}
+		inlen = len;
+	}
+
+	if (!hashdata->ctx.data) {
+		retstatus = do_allocate_intern(hashdata);
+		if (retstatus != CAAM_NO_ERROR) {
+			ret = TEE_ERROR_OUT_OF_MEMORY;
+			goto exit_update;
+		}
+	}
+
+	HASH_TRACE("Update Type 0x%X - Input @0x%p-%zu", alg->type,
```

Comment:
s/%X/%"PRIx32"/

minor: remove `0x` in front of `%p`.

Reply:
ok